security/security.c

```diff
@@ -464,6 +469,45 @@ static int lsm_append(const char *new, char **result)
 	return 0;
 }
 
+#define LSM_FUNC_DEFAULT(NAME)		NAME##_func_default
+#define HOOK_STATIC_CALL(HOOK, NUM)	static_call_##HOOK##_##NUM
+#define HOOK_STATIC_CHECK(HOOK, NUM)	static_check_##HOOK##_##NUM
+
+#define FOR_EACH_HOOK_SLOT(M, ...)			\
+	M(1, __VA_ARGS__)				\
+	M(2, __VA_ARGS__)				\
+	M(3, __VA_ARGS__)
+
+#define CREATE_STATIC(NUM, NAME)			\
+	DEFINE_STATIC_CALL(HOOK_STATIC_CALL(NAME, NUM), LSM_FUNC_DEFAULT(NAME));	\
+	DEFINE_STATIC_KEY_FALSE(HOOK_STATIC_CHECK(NAME, NUM));
+
+// We need a default function that will not be called so that 
+// static_call can infer the expected type
+#define LSM_HOOK(RET, DEFAULT, NAME, ...) 		\
+	noinline RET LSM_FUNC_DEFAULT(NAME)(__VA_ARGS__)\
+	{						\
+		return DEFAULT;				\
+	}						\
+	FOR_EACH_HOOK_SLOT(CREATE_STATIC, NAME)
+#include <linux/lsm_hook_defs.h>
+#undef LSM_HOOK
+#undef CREATE_STATIC
+
+#define TRY_TO_ADD(NUM, HOOK, FUNC)			\
+	if (!static_branch_unlikely(&HOOK_STATIC_CHECK(HOOK, NUM))) {	\
+		static_call_update(HOOK_STATIC_CALL(HOOK, NUM), FUNC); 	\
+		static_branch_enable(&HOOK_STATIC_CHECK(HOOK, NUM)); 	\
+		break;							\
+	}
+
+#define add_static_hook(HOOK, FUNC)			\
+	do {						\
+		FOR_EACH_HOOK_SLOT(TRY_TO_ADD, HOOK, FUNC)		\
+		printk(KERN_ERR "No slot remaining to add LSM hook for "\
+		       #HOOK "\n"); \
+	} while(0)
```

Comment:
It's gonna be quite fiddly, but I think instead of having a globally fixed number of slots, and trying to put each hook into a slot one by one, we should try to:
- determine the number of LSMs that exist at compile time, and set the number of slots to that-
- at boot time, fix the mapping from LSMs to slot numbers.